docs/design-documents/drivers/serial/serial.md

```diff
@@ -0,0 +1,321 @@
+# Introduction
+
+### Overview and background
+
+Mbed OS contains multiple serial classes and multiple ways of printing information to the console. Serial classes have evolved over the years to add new functinality to existing classes and compensate for limitations. The result is that there are now multiple variants, not always well documented; therefore it is not clear which class shoud be used for what purpose. Futhermore, some of the classes pull substantial dependencies whose benefits are not very clear, for example standard library like objects and abstractions, and parts of the retarget layer.
+
+This document collects proposals made by various contributors (see [references](#references)) and presents a simplified serial class hierarchy. The aim is to offer two public classes, one for unbuffered I/O and the other for buffered I/O, with a clear definition of which one should be used when. All other serial classes are deprecated and should be removed at a later stage. In addition, the serial classes dependencies will be optimised to only include components that provide valuable functionality.
+
+#### Serial classes in Mbed OS 5.14
+
+**Serial**
+
+`Serial` provides unbuffered I/O and is using blocking HAL calls. It only works for simple printing. 
+
+Its main limitations are:
+- It uses a mutex lock so it cannot be used from interrupts. 
+- It wastes a lot of CPU time spinning waiting for serial output and it cannot be used reliably for input because it needs interrupt-speed response on read calls but its high-level API cannot be used from interrupts.
+- It lacks buffering so it cannot be used reliably for input and output without flow control.
+- It pulls in the C library stdio system because it uses `Stream`.
+
+**UARTSerial**
+
+`UARTSerial` provides buffered I/O. It can be used reliably for input from non-interrupt context and to avoid excess spinning waiting for transmission buffer space. It also does better blocking and correct POSIX-like `FileHandle` semantics. 
+
+It does not use `Stream` so it has the advantage of not pulling in the C stdio library. The drawback is that it loses the ability to call printf directly.
+
+**RawSerial**
+
+`RawSerial` provides unbuffered I/O. It has no locks so it is safe to use in interrupts provided that it is the only instance that is using that serial port.
+
+### Requirements and assumptions
+
+None
+
+# System architecture and high-level design
+
+Applications will have a choice between two public serial classes: 
+- `BufferedSerial` for a normal application
+- `UnbufferedSerial` for an interrupt driven application or one that needs to have more control
+
+### System architecture and component interaction
+
+The below diagram shows the inheritance hierarchy for the serial classes.
+
+![Serial classes](./diagrams/serial-classes.png)
+
+`MinimalSerial` is an abstract class that extracts basics I/O functionality from `SerialBase` so it can be re-used for a minimal serial console in the retarget code.
```

Comment:
I'm a bit doubtful about the value of this, as a separation, but will reserve judgement until I see it.

Reply:
Since this is what is used by the minimal console in the simplified retarget code, it should save a bit of memory and make the MinimalConsole implementation simpler. But if this is not the case, we will revert.

Reply:
Using `DirectSerial`
```
Total Static RAM memory (data + bss): 205934 bytes
Total Flash memory (text + data): 46299 bytes
```

The below are with `MinimalConsole` inheriting from `MinimalSerial`
```
Total Static RAM memory (data + bss): 205978 bytes
Total Flash memory (text + data): 44352 bytes
```

The below are with `MinimalConsole` without inheritance
```
Total Static RAM memory (data + bss): 205782(-196) bytes
Total Flash memory (text + data): 44312(-40) bytes
```

Reply:
MinimalSerial is not bringing memory ssaving so it is removed.